OrbitProducer/CaptureEventProducer.cpp

```diff
@@ -0,0 +1,203 @@
+}
+
+void CaptureEventProducer::OnCaptureStart() { LOG("CaptureEventProducer called OnCaptureStart"); }
+
+void CaptureEventProducer::OnCaptureStop() { LOG("CaptureEventProducer called OnCaptureStop"); }
+
+bool CaptureEventProducer::SendCaptureEvents(
+    const orbit_grpc_protos::ReceiveCommandsAndSendEventsRequest& send_events_request) {
+  CHECK(send_events_request.event_case() ==
+        orbit_grpc_protos::ReceiveCommandsAndSendEventsRequest::kBufferedCaptureEvents);
+
+  CHECK(producer_side_service_stub_ != nullptr);
+  {
+    // Acquiring the mutex just for the CHECK might seem expensive,
+    // but the gRPC call that follows is orders of magnitude slower.
+    absl::ReaderMutexLock lock{&shutdown_requested_mutex_};
+    CHECK(!shutdown_requested_);
+  }
+
+  bool write_succeeded;
+  {
+    absl::ReaderMutexLock lock{&context_and_stream_mutex_};
+    if (stream_ == nullptr) {
+      ERROR("Sending BufferedCaptureEvents to ProducerSideService: not connected");
+      return false;
+    }
+    write_succeeded = stream_->Write(send_events_request);
+  }
+  if (!write_succeeded) {
+    ERROR("Sending BufferedCaptureEvents to ProducerSideService");
+  }
+  return write_succeeded;
+}
+
+bool CaptureEventProducer::NotifyAllEventsSent() {
+  CHECK(producer_side_service_stub_ != nullptr);
+  {
+    absl::ReaderMutexLock lock{&shutdown_requested_mutex_};
+    CHECK(!shutdown_requested_);
+  }
+
+  orbit_grpc_protos::ReceiveCommandsAndSendEventsRequest all_events_sent_request;
+  all_events_sent_request.mutable_all_events_sent();
+  bool write_succeeded;
+  {
+    absl::ReaderMutexLock lock{&context_and_stream_mutex_};
+    if (stream_ == nullptr) {
+      ERROR("Sending AllEventsSent to ProducerSideService: not connected");
+      return false;
+    }
+    write_succeeded = stream_->Write(all_events_sent_request);
+  }
+  if (write_succeeded) {
+    LOG("Sent AllEventsSent to ProducerSideService");
+  } else {
+    ERROR("Sending AllEventsSent to ProducerSideService");
+  }
+  return write_succeeded;
+}
+
+void CaptureEventProducer::ConnectAndReceiveCommandsThread() {
+  CHECK(producer_side_service_stub_ != nullptr);
+
+  while (true) {
+    {
+      absl::ReaderMutexLock lock{&shutdown_requested_mutex_};
+      if (shutdown_requested_) {
+        break;
+      }
+    }
+
+    // Attempt to connect to ProducerSideService. Note that getting a stream_ != nullptr doesn't
+    // mean that the service is listening nor that the connection is actually established.
+    {
+      absl::ReaderMutexLock lock{&context_and_stream_mutex_};
+      context_ = std::make_unique<grpc::ClientContext>();
+      stream_ = producer_side_service_stub_->ReceiveCommandsAndSendEvents(context_.get());
+    }
+
+    if (stream_ == nullptr) {
+      ERROR(
+          "Calling ReceiveCommandsAndSendEvents to establish "
+          "gRPC connection with ProducerSideService");
+      // This is the reason why we protect shutdown_requested_ with a Mutex instead of
```

Comment:
Nit: Mutex -> mutex

Reply:
This is intended, as I'm not referring to any mutex, but to `absl::Mutex`, which supports `LockWhenWithTimeout`. Added `absl::` in the comment.